docs/design/coreclr/botr/memory-model.md

```diff
@@ -0,0 +1,248 @@
+
+## Alignment
+When managed by CLR runtime, variables of built-in primitive types are *properly aligned* according to the data type size. This applies to both heap and stack allocated memory.
+
+1-byte, 2-byte, 4-byte variables are stored at 1-byte, 2-byte, 4-byte boundary, respectively.
+8-byte variables are 8-byte aligned on 64 bit platforms.
+Native-sized integer types and pointers have alignment that matches their size on the given platform.
+
+## Atomic memory accesses.
+Memory accesses to *properly aligned* data of primitive types are always atomic. The value that is observed is always a result of complete read and write operations.
+
+## Unmanaged memory access.
+As unmanaged pointers can point to any addressable memory, operations with such pointers may violate guarantees provided by the runtime and expose undefined or platform-specific behavior.
+**Example:** memory accesses through pointers which are *not properly aligned* may be not atomic or cause faults depending on the platform and hardware configuration.
+
+Although rare, unaligned access is a realistic scenario and thus there is some limited support for unaligned memory accesses, such as:
+* `.unaligned` IL prefix
+* `Unsafe.ReadUnaligned`, `Unsafe.WriteUnaligned` and ` Unsafe.CopyBlockUnaligned` helpers.
+
+These facilities ensure fault-free access to potentially unaligned locations, but do not ensure atomicity.
+
+As of this writing there is no specific support for operating with incoherent memory, device memory or similar. Passing non-ordinary memory to the runtime by the means of pointer operations or native interop results in Undefined Behavior.
+
+## Sideeffects and optimizations of memory accesses.
+CLR assumes that the sideeffects of memory reads and writes include only changing and observing values at specified memory locations. This applies to all reads and writes - volatile or not. **This is different from ECMA model.**
+
+As a consequence:
+* Speculative writes are not allowed.
+* Reads cannot be introduced.
+* Unused reads can be elided.
+* Adjacent nonvolatile reads from the same location can be coalesced.
+* Adjacent nonvolatile writes to the same location can be coalesced.
+
+## Thread-local memory accesses.
+It may be possible for an optimizing compiler to prove that some data is accessible only by a single thread. In such case it is permitted to perform further optimizations such as duplicating or removal of memory accesses.
+
+## Cross-thread access to local variables.
+-	There is no type-safe mechanism for accessing locations on one thread’s stack from another thread.
+-	Accessing managed references located on the stack of a different thread by the means of unsafe code will result in Undefined Behavior.
```

Comment:
I am curious -- what is an example of UB one can observe today when passing pointers to locals around threads?

Reply:
There is an obvious hazard form the other stack getting shorter and then longer and recycling the location for something completely else.
Then runtime (JIT and GC primarily) make assumptions that the stack is not shared. That allows important optimizations for single-threaded access to objects, which is a common case. For example assigning to local variables is not reported to GC (does not perform GC-write-barriers) as assigning to a non-shared location cannot change the object's generation, cannot expose it to other threads prematurely, etc... 

As an example that does not need a lot of stress to fail. The following would typically crash on ARM64 in under 1 second: 
```cs
internal unsafe class Program
{
  private static void Main(string[] args)
  {
    Task.Run(M1);
    while (ptrStatic == null) { Thread.Sleep(1);}

    ref object otherThreadsLocal = ref Unsafe.AsRef<object>(ptrStatic);
    
    for(; ; )
    {
      System.Console.WriteLine(otherThreadsLocal); // can see corrupted incomplete objects
    }
  }

  static void* ptrStatic;

  static void M1()
  {   
    object local1 = 0;
    ptrStatic = Unsafe.AsPointer(ref local1);

    for (long i = 0; i < long.MaxValue; i++)
    {
        local1 = i; 
    }
  }
}

```